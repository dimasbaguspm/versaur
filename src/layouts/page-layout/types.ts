import type { HTMLAttributes } from 'react'

/**
 * PageLayoutRoot - main container for page layout
 * Manages layout between header and content regions
 * Compound pattern: HeaderRegion, ContentRegion
 */
export interface PageLayoutProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Add horizontal margin (left and right) to the page layout
   * @default false
   */
  hasMargin?: boolean
}

/**
 * PageLayoutHeaderRegion - header region of the page
 */
export interface PageLayoutHeaderRegionProps
  extends HTMLAttributes<HTMLDivElement> {
  /**
   * Background color of the header region
   * - 'white': White background
   * - 'gray': Gray background
   * @default 'white'
   */
  backgroundColor?: 'white' | 'gray'
}

/**
 * PageLayoutContentRegion - content region of the page
 */
export interface PageLayoutContentRegionProps
  extends HTMLAttributes<HTMLDivElement> {
  /**
   * Background color of the content region
   * - 'white': White background
   * - 'gray': Gray background
   * @default 'white'
   */
  backgroundColor?: 'white' | 'gray'
}
