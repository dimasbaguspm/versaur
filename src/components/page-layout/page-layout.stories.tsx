/**
 * PageLayout stories
 *
 * @group Layout
 * @description
 * The PageLayout component provides a consistent responsive container for Versaur apps.
 * Use the `type` prop to switch between desktop, tablet, and mobile breakpoints.
 */
import type { Meta, StoryObj } from '@storybook/react'
import { PageLayout } from './page-layout'

const meta: Meta<typeof PageLayout> = {
  title: 'Layout/PageLayout',
  component: PageLayout,
  tags: ['autodocs'],
  parameters: {
    layout: 'fullscreen',
  },
  argTypes: {
    type: {
      control: 'radio',
      options: ['desktop', 'tablet', 'mobile'],
      description: 'Breakpoint type for layout',
      defaultValue: 'desktop',
    },
    className: { control: false },
    children: { control: false },
  },
}
export default meta

type Story = StoryObj<typeof PageLayout>

export const Desktop: Story = {
  args: {
    type: 'desktop',
    children: 'Desktop layout (max-w-7xl)',
  },
}

export const Tablet: Story = {
  args: {
    type: 'tablet',
    children: 'Tablet layout (max-w-3xl)',
  },
}

export const Mobile: Story = {
  args: {
    type: 'mobile',
    children: 'Mobile layout (full width)',
  },
}
