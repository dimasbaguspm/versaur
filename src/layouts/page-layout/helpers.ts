import { cva } from 'class-variance-authority'
/**
 * Tailwind variants for PageLayout
 */

export const pageLayoutVariants = cva('w-full mx-auto relative', {
  variants: {
    type: {
      desktop: 'max-w-7xl px-6 pb-10',
      tablet: 'max-w-3xl px-4',
      mobile: 'w-full px-4',
    },
  },
  defaultVariants: {
    type: 'desktop',
  },
})
