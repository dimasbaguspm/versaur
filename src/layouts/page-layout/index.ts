export { PageLayout } from './page-layout'
export type * from './types'
