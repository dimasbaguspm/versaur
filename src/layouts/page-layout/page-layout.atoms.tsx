import { forwardRef } from 'react'
import {
  pageLayoutHeaderRegionStyles,
  pageLayoutContentRegionStyles,
} from './helpers'
import type {
  PageLayoutHeaderRegionProps,
  PageLayoutContentRegionProps,
} from './types'

export const PageLayoutHeaderRegion = forwardRef<
  HTMLDivElement,
  PageLayoutHeaderRegionProps
>(function PageLayoutHeaderRegion(
  { children, className, backgroundColor = 'white', ...props },
  ref
) {
  return (
    <div
      ref={ref}
      className={pageLayoutHeaderRegionStyles({ backgroundColor, className })}
      {...props}
    >
      {children}
    </div>
  )
})

export const PageLayoutContentRegion = forwardRef<
  HTMLDivElement,
  PageLayoutContentRegionProps
>(function PageLayoutContentRegion(
  { children, className, backgroundColor = 'white', ...props },
  ref
) {
  return (
    <div
      ref={ref}
      className={pageLayoutContentRegionStyles({ backgroundColor, className })}
      {...props}
    >
      {children}
    </div>
  )
})
