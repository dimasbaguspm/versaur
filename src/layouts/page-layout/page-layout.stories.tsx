/**
 * PageLayout stories
 *
 * @group Layout
 * @description
 * The PageLayout component serves as a parent wrapper for main content,
 * sitting in the body region or below the page header. It provides flexible
 * sizing, column templates, and background color options.
 */
import type { Meta, StoryObj } from '@storybook/react'
import { PageLayout } from './page-layout'

const meta: Meta<typeof PageLayout> = {
  title: 'Layouts/PageLayout',
  component: PageLayout,
  tags: ['autodocs'],
  parameters: {
    layout: 'fullscreen',
  },
  argTypes: {
    size: {
      control: 'radio',
      options: ['fluid', 'wide', 'narrow'],
    },
    template: {
      control: 'radio',
      options: [
        'single-column',
        'two-column',
        'two-column-asymmetric-left',
        'two-column-asymmetric-right',
      ],
    },
    backgroundColor: {
      control: 'radio',
      options: ['white', 'gray'],
    },
  },
  decorators: [
    Story => (
      <div className='min-h-screen'>
        <Story />
      </div>
    ),
  ],
}
export default meta

type Story = StoryObj<typeof PageLayout>

const SampleContent = ({ label = 'Content' }: { label?: string }) => (
  <div className='rounded-lg bg-primary-soft p-6 text-foreground'>
    <h2 className='mb-2 text-xl font-semibold'>{label}</h2>
    <p className='text-foreground-light'>
      This is sample content to demonstrate the layout behavior.
    </p>
  </div>
)

/**
 * Default PageLayout with fluid size, single column, and white background
 */
export const Default: Story = {
  args: {
    children: <SampleContent />,
  },
}

/**
 * Wide container with centered content (desktop screens)
 */
export const WideContainer: Story = {
  args: {
    size: 'wide',
    children: <SampleContent label='Wide Container Content' />,
  },
}

/**
 * Narrow container with centered content (mobile screens)
 */
export const NarrowContainer: Story = {
  args: {
    size: 'narrow',
    children: <SampleContent label='Narrow Container Content' />,
  },
}

/**
 * Two equal columns layout
 */
export const TwoColumn: Story = {
  args: {
    size: 'wide',
    template: 'two-column',
    children: (
      <>
        <SampleContent label='Column 1' />
        <SampleContent label='Column 2' />
      </>
    ),
  },
}

/**
 * Two columns with left column wider (2:1 ratio)
 */
export const TwoColumnAsymmetricLeft: Story = {
  args: {
    size: 'wide',
    template: 'two-column-asymmetric-left',
    children: (
      <>
        <div className='md:col-span-2'>
          <SampleContent label='Main Content (Wider)' />
        </div>
        <SampleContent label='Sidebar' />
      </>
    ),
  },
}

/**
 * Two columns with right column wider (1:2 ratio)
 */
export const TwoColumnAsymmetricRight: Story = {
  args: {
    size: 'wide',
    template: 'two-column-asymmetric-right',
    children: (
      <>
        <SampleContent label='Sidebar' />
        <div className='md:col-span-2'>
          <SampleContent label='Main Content (Wider)' />
        </div>
      </>
    ),
  },
}

/**
 * Gray background for visual contrast
 */
export const GrayBackground: Story = {
  args: {
    size: 'wide',
    backgroundColor: 'gray',
    children: (
      <div className='rounded-lg bg-white p-6'>
        <h2 className='mb-2 text-xl font-semibold'>
          Content on Gray Background
        </h2>
        <p className='text-foreground-light'>
          The gray background provides visual contrast for the page.
        </p>
      </div>
    ),
  },
}
