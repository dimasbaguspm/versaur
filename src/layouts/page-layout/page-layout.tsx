import { forwardRef } from 'react'
import type { PageLayoutProps } from './types'
import { pageLayoutRootStyles } from './helpers'

import {
  PageLayoutHeaderRegion,
  PageLayoutContentRegion,
} from './page-layout.atoms'

/**
 * PageLayout - Compound layout component for page-level structure
 *
 * Provides a layout system for managing the structure between page header and content:
 * - HeaderRegion: Page header area (PageHeader component)
 * - ContentRegion: Main content area (PageContent component)
 *
 * Uses CSS Grid with grid-area for robust layout management.
 * Takes responsibility for backgroundColor from PageHeader and PageContent.
 *
 * @example
 * ```tsx
 * <PageLayout hasMargin>
 *   <PageLayout.HeaderRegion backgroundColor="white">
 *     <PageHeader>...</PageHeader>
 *   </PageLayout.HeaderRegion>
 *   <PageLayout.ContentRegion backgroundColor="gray">
 *     <PageContent>...</PageContent>
 *   </PageLayout.ContentRegion>
 * </PageLayout>
 * ```
 */
const PageLayoutRoot = forwardRef<HTMLDivElement, PageLayoutProps>(
  ({ children, className, hasMargin = false, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={pageLayoutRootStyles({ hasMargin, className })}
        {...props}
      >
        {children}
      </div>
    )
  }
)

export const PageLayout = Object.assign(PageLayoutRoot, {
  HeaderRegion: PageLayoutHeaderRegion,
  ContentRegion: PageLayoutContentRegion,
})
