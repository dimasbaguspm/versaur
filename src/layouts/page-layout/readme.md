# PageLayout

A compound layout component for managing page-level structure between header and content regions.

## Installation

```bash
npm install versaur
```

## Usage

```tsx
import { PageLayout } from 'versaur/layouts'
import { PageHeader, PageContent } from 'versaur/layouts'

function MyPage() {
  return (
    <PageLayout>
      <PageLayout.HeaderRegion>
        <PageHeader title='Page Title' subtitle='Description' />
      </PageLayout.HeaderRegion>
      <PageLayout.ContentRegion>
        <PageContent size='wide'>{/* Your content here */}</PageContent>
      </PageLayout.ContentRegion>
    </PageLayout>
  )
}
```

## Props

### PageLayout (Root)

| Prop        | Type      | Default | Description                               |
| ----------- | --------- | ------- | ----------------------------------------- |
| `hasMargin` | `boolean` | `false` | Add horizontal margins to the page layout |

### HeaderRegion

| Prop              | Type                | Default   | Description                           |
| ----------------- | ------------------- | --------- | ------------------------------------- |
| `backgroundColor` | `'white' \| 'gray'` | `'white'` | Background color of the header region |

### ContentRegion

| Prop              | Type                | Default   | Description                            |
| ----------------- | ------------------- | --------- | -------------------------------------- |
| `backgroundColor` | `'white' \| 'gray'` | `'white'` | Background color of the content region |

## Examples

### With Margins

```tsx
<PageLayout hasMargin>
  <PageLayout.HeaderRegion>
    <PageHeader title='Page Title' />
  </PageLayout.HeaderRegion>
  <PageLayout.ContentRegion>
    <PageContent size='wide'>{/* content */}</PageContent>
  </PageLayout.ContentRegion>
</PageLayout>
```

### Different Backgrounds

```tsx
<PageLayout>
  <PageLayout.HeaderRegion backgroundColor='gray'>
    <PageHeader title='Dashboard' />
  </PageLayout.HeaderRegion>
  <PageLayout.ContentRegion backgroundColor='white'>
    <PageContent size='wide'>{/* content */}</PageContent>
  </PageLayout.ContentRegion>
</PageLayout>
```

## See Also

- [PageHeader](/layouts/page-header)
- [PageContent](/layouts/page-content)
- [AppLayout](/layouts/app-layout)
