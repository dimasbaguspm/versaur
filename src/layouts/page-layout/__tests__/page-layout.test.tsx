import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../page-layout.stories'

describe('PageLayout', () => {
  const {
    Basic,
    WithMargins,
    GrayHeaderWhiteContent,
    WhiteHeaderGrayContent,
    BothGray,
    NarrowLayout,
    FluidLayout,
  } = composeStories(stories)

  describe('Basic story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<Basic />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders header region with page header', () => {
      render(<Basic />)
      expect(screen.getByText('Page Title')).toBeInTheDocument()
      expect(
        screen.getByText('A brief description of this page')
      ).toBeInTheDocument()
    })

    it('renders content region with scrollable content', () => {
      render(<Basic />)
      expect(screen.getByText('Section 1')).toBeInTheDocument()
      expect(screen.getByText('Section 5')).toBeInTheDocument()
    })
  })

  describe('WithMargins story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<WithMargins />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('applies horizontal margins when hasMargin is true', () => {
      const { container } = render(<WithMargins />)
      const root = container.firstChild
      expect(root).toHaveClass('[&_[class*="grid-area:header"]]:px-4')
    })

    it('renders page title', () => {
      render(<WithMargins />)
      expect(screen.getByText('Page with Margins')).toBeInTheDocument()
    })
  })

  describe('GrayHeaderWhiteContent story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<GrayHeaderWhiteContent />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders header with gray background', () => {
      const { container } = render(<GrayHeaderWhiteContent />)
      const headerRegion = container.querySelector(
        '[class*="grid-area:header"]'
      )
      expect(headerRegion).toHaveClass('bg-gray-100')
    })

    it('renders badges', () => {
      render(<GrayHeaderWhiteContent />)
      expect(screen.getByText('Active')).toBeInTheDocument()
      expect(screen.getByText('Beta')).toBeInTheDocument()
    })

    it('renders action button', () => {
      render(<GrayHeaderWhiteContent />)
      expect(screen.getByText('Add Widget')).toBeInTheDocument()
    })
  })

  describe('WhiteHeaderGrayContent story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<WhiteHeaderGrayContent />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders content with gray background', () => {
      const { container } = render(<WhiteHeaderGrayContent />)
      const contentRegion = container.querySelector(
        '[class*="grid-area:content"]'
      )
      expect(contentRegion).toHaveClass('bg-gray-100')
    })
  })

  describe('BothGray story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<BothGray />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('applies gray background to both regions', () => {
      const { container } = render(<BothGray />)
      const regions = container.querySelectorAll('[class*="bg-gray-100"]')
      expect(regions.length).toBeGreaterThan(0)
    })

    it('renders reports content', () => {
      render(<BothGray />)
      expect(screen.getByText('Reports')).toBeInTheDocument()
      expect(screen.getByText('Report 1')).toBeInTheDocument()
      expect(screen.getByText('Report 5')).toBeInTheDocument()
    })
  })

  describe('NarrowLayout story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<NarrowLayout />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders narrow layout with margins', () => {
      const { container } = render(<NarrowLayout />)
      const root = container.firstChild
      expect(root).toHaveClass('[&_[class*="grid-area:header"]]:px-4')
    })

    it('renders article content', () => {
      render(<NarrowLayout />)
      expect(screen.getByText('Article Title')).toBeInTheDocument()
      expect(screen.getByText('Paragraph 1')).toBeInTheDocument()
      expect(screen.getByText('Paragraph 8')).toBeInTheDocument()
    })
  })

  describe('FluidLayout story', () => {
    it('renders correctly (snapshot)', () => {
      const { asFragment } = render(<FluidLayout />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders full-width dashboard', () => {
      render(<FluidLayout />)
      expect(screen.getByText('Full Width Dashboard')).toBeInTheDocument()
      expect(screen.getByText('Panel 1')).toBeInTheDocument()
      expect(screen.getByText('Panel 5')).toBeInTheDocument()
    })
  })

  describe('Accessibility', () => {
    it('uses semantic HTML grid structure', () => {
      const { container } = render(<Basic />)
      const root = container.firstChild as HTMLElement
      expect(root).toHaveClass('grid')
    })

    it('applies proper grid areas for regions', () => {
      const { container } = render(<GrayHeaderWhiteContent />)
      const headerRegion = container.querySelector(
        '[class*="grid-area:header"]'
      )
      const contentRegion = container.querySelector(
        '[class*="grid-area:content"]'
      )
      expect(headerRegion).toBeInTheDocument()
      expect(contentRegion).toBeInTheDocument()
    })

    it('supports keyboard navigation through content', () => {
      render(<Basic />)
      const content = screen.getByText('Section 1')
      expect(content).toBeInTheDocument()
    })
  })

  describe('Layout behavior', () => {
    it('applies grid template areas correctly', () => {
      const { container } = render(<Basic />)
      const root = container.firstChild as HTMLElement
      expect(root).toHaveClass('grid-rows-[auto_1fr]')
    })

    it('makes content region scrollable', () => {
      const { container } = render(<Basic />)
      const contentRegion = container.querySelector(
        '[class*="grid-area:content"]'
      )
      expect(contentRegion).toHaveClass('overflow-y-auto')
    })

    it('does not apply margins by default', () => {
      const { container } = render(<Basic />)
      const root = container.firstChild as HTMLElement
      expect(root).not.toHaveClass('mx-4')
    })
  })
})
