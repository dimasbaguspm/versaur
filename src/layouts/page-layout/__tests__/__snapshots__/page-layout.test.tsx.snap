// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageLayout > renders desktop layout and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full mx-auto relative max-w-7xl px-6 pb-10"
  >
    Desktop layout (max-w-7xl)
  </div>
</DocumentFragment>
`;

exports[`PageLayout > renders mobile layout and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mx-auto relative w-full px-4"
  >
    Mobile layout (full width)
  </div>
</DocumentFragment>
`;

exports[`PageLayout > renders tablet layout and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full mx-auto relative max-w-3xl px-4"
  >
    Tablet layout (max-w-3xl)
  </div>
</DocumentFragment>
`;
